common-lib/src/main/scala/com/gu/mediaservice/lib/config/ImageProcessorLoader.scala

```diff
@@ -0,0 +1,109 @@
+    val configDetails = parseConfigValue(config.getValue(path), path)
+    loadImageProcessor(configDetails)
+  }
+
+  private def parseConfigValue(configValue: ConfigValue, path: String): ImageProcessorConfigDetails = {
+    configValue match {
+      case plainClass if plainClass.valueType == ConfigValueType.STRING =>
+        ImageProcessorConfigDetails(plainClass.unwrapped.asInstanceOf[String], None, plainClass.origin, path)
+      case withConfig:ConfigObject if validConfigObject(withConfig) =>
+        val config = withConfig.toConfig
+        val className = config.getString("className")
+        val processorConfig = config.getConfig("config")
+        ImageProcessorConfigDetails(className, Some(Configuration(processorConfig)), withConfig.origin, path)
+      case _ =>
+        throw new BadValue(configValue.origin, path, s"An image processor can either a class name (string) or object with className (string) and config (object) fields. This ${configValue.valueType} is not valid.")
+    }
+  }
+
+  private def validConfigObject(configObject: ConfigObject): Boolean = {
+    val config = configObject.toConfig
+    config.hasPath("className") && config.hasPath("config")
+  }
+
+  private def loadImageProcessor(details: ImageProcessorConfigDetails): ImageProcessor = {
+    ImageProcessorLoader
+      .loadImageProcessor(details.className, details.config.getOrElse(Configuration.empty))
+      .getOrElse { error =>
+        val configError = s"Unable to instantiate image processor from config: $error"
+        logger.error(configError)
+        throw new BadValue(details.origin, details.path, configError)
+      }
+  }
+
+  def loadImageProcessor(className: String, config: Configuration): Either[String, ImageProcessor] = {
+    for {
+      imageProcessorClass <- loadClass(className)
+      imageProcessorInstance <- instantiate(imageProcessorClass, config)
+    } yield imageProcessorInstance
+  }
+
+  private def loadClass(className: String): Either[String, Class[_]] = catchNonFatal(Class.forName(className)) {
+    case _: ClassNotFoundException => s"Unable to find image processor class $className"
+    case other =>
+      logger.error(s"Error whilst loading $className", other)
+      s"Unknown error whilst loading $className, check logs"
+  }
+
+  private def instantiate(imageProcessorClass: Class[_], config: Configuration): Either[String, ImageProcessor] = {
+    // Fail fast if config is provided but the specified image processor ignores it
+    def assertNoConfiguration[T](ok: Right[String, T]): Either[String, T] = {
+      if (config.keys.nonEmpty) {
+        Left(s"Attempt to initialise image processor ${imageProcessorClass.getCanonicalName} failed as configuration is provided but the constructor does not take configuration as an argument.")
+      } else {
+        ok
+      }
+    }
+
+    val maybeCompanionObject = Try(imageProcessorClass.getField("MODULE$").get(imageProcessorClass)).toOption
```

Comment:
Do we need to support companion objects? I'm worried this would break in future versions of Scala

Reply:
I originally didn't but a lot of the existing code is objects and actually that is what they should be if the processor is pure. I wrote a test for all of these cases so if it does break we'll notice swiftly. In that case we can either fix it to use the scala reflection or disable support.